File: windows-client/astral_client/streaming.py
"""Converts orchestrator ui_stream_data frames into Canvas ops, rendering only the
structured components (never the web html). Keys nodes by stream_id, or component_id
when bridged to a workspace identity; used by astral_client/app.py.
"""

from __future__ import annotations

from typing import Container, Dict, List, Optional, Tuple

STREAM_NODE_PREFIX = "stream-"


def stream_node_id(stream_id: str) -> str:
    return f"{STREAM_NODE_PREFIX}{stream_id}"


def _node_key(frame: dict) -> Tuple[Optional[str], Optional[str]]:
    sid = frame.get("stream_id")
    if sid:
        cid = frame.get("component_id")
        return (str(cid) if cid else stream_node_id(str(sid))), str(sid)
    tool = frame.get("tool_name")
    if tool:
        return f"{STREAM_NODE_PREFIX}tool-{tool}", f"tool:{tool}"
    return None, None


def _error_component(err: dict) -> dict:
    retryable = bool(err.get("retryable"))
    text = err.get("message") or err.get("code") or "stream error"
    return {
        "type": "alert",
        "variant": "warning" if retryable else "error",
        "title": "Live update interrupted" if retryable else "Live update failed",
        "message": str(text),
    }


def stream_frame_to_ops(
    frame: dict, *, active_chat: Optional[str], seq_state: Dict[str, int]
) -> List[dict]:
    node, key = _node_key(frame)
    if not node:
        return []

    session = frame.get("session_id")
    if session and active_chat and session != active_chat:
        return []

    seq = frame.get("seq")
    if isinstance(seq, int) and key is not None:
        last = seq_state.get(key)
        if last is not None and seq <= last:
            return []
        seq_state[key] = seq

    if frame.get("terminal") and key is not None:
        seq_state.pop(key, None)

    err = frame.get("error")
    if isinstance(err, dict) and err:
        return [{"op": "upsert", "component_id": node, "component": _error_component(err)}]

    comps = [c for c in (frame.get("components") or []) if isinstance(c, dict)]
    if not comps:
        return []
    body = comps[0] if len(comps) == 1 else {"type": "container", "content": comps}
    return [{"op": "upsert", "component_id": node, "component": body}]


def subscribe_ack_ops(frame: dict, *, existing_ids: Container[str] = ()) -> List[dict]:
    node, _ = _node_key(frame)
    if not node or node in existing_ids:
        return []
    tool = frame.get("tool_name") or "tool"
    return [{
        "op": "upsert",
        "component_id": node,
        "component": {"type": "text", "content": f"Streaming {tool}…"},
    }]


def stream_error_ops(frame: dict) -> List[dict]:
    payload = frame.get("payload") or {}
    node, _ = _node_key({
        "stream_id": payload.get("stream_id"),
        "tool_name": payload.get("tool_name") or frame.get("tool_name"),
    })
    if not node:
        return []
    text = payload.get("message") or payload.get("code") or frame.get("error") or "stream error"
    return [{
        "op": "upsert",
        "component_id": node,
        "component": {"type": "alert", "variant": "error", "title": "Stream error", "message": str(text)},
    }]

File: android-client/core/src/main/kotlin/com/personalailabs/astraldeep/core/streaming/Streaming.kt
// Translates the orchestrator's push-streaming protocol frames into Canvas ops, porting the verified Windows
// streaming.py; nodeKey bridges a stream to its workspace component_id when present, else a synthetic node.

package com.personalailabs.astraldeep.core.streaming

import com.personalailabs.astraldeep.core.protocol.Inbound
import com.personalailabs.astraldeep.core.protocol.StreamError
import com.personalailabs.astraldeep.core.sdui.CanvasOp
import com.personalailabs.astraldeep.core.sdui.Component
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.put

const val STREAM_NODE_PREFIX = "stream-"

fun streamNodeId(streamId: String): String = "$STREAM_NODE_PREFIX$streamId"

private fun nodeKey(
    streamId: String?,
    toolName: String?,
    componentId: String? = null,
): Pair<String, String>? =
    when {
        streamId != null -> (componentId ?: streamNodeId(streamId)) to streamId
        toolName != null -> "${STREAM_NODE_PREFIX}tool-$toolName" to "tool:$toolName"
        else -> null
    }

private fun errorComponent(
    node: String,
    err: StreamError,
): Component {
    val retryable = err.retryable
    val text = err.message ?: err.code ?: "stream error"
    val attrs =
        buildJsonObject {
            put("type", "alert")
            put("variant", if (retryable) "warning" else "error")
            put("title", if (retryable) "Live update interrupted" else "Live update failed")
            put("message", text)
        }
    return Component(type = "alert", id = node, attributes = attrs, children = emptyList())
}

private fun containerOf(
    node: String,
    comps: List<Component>,
): Component =
    Component(
        type = "container",
        id = node,
        attributes = buildJsonObject { put("type", "container") },
        children = comps,
    )

fun streamFrameToOps(
    frame: Inbound.UiStreamData,
    activeChat: String?,
    seqState: MutableMap<String, Int>,
): List<CanvasOp> {
    val (node, key) = nodeKey(frame.streamId, frame.toolName, frame.componentId) ?: return emptyList()

    val session = frame.sessionId
    if (session != null && activeChat != null && session != activeChat) return emptyList()

    val seq = frame.seq
    if (seq != null) {
        val last = seqState[key]
        if (last != null && seq <= last) return emptyList()
        seqState[key] = seq
    }
    if (frame.terminal) seqState.remove(key)

    frame.error?.let { return listOf(CanvasOp("upsert", node, errorComponent(node, it))) }

    val comps = frame.components
    if (comps.isEmpty()) return emptyList()
    val body = if (comps.size == 1) comps[0].copy(id = node) else containerOf(node, comps)
    return listOf(CanvasOp("upsert", node, body))
}

// Guards late joiners: must not blank content already retained by id
fun subscribeAckOps(
    msg: Inbound.StreamSubscribed,
    existingIds: Set<String> = emptySet(),
): List<CanvasOp> {
    val (node, _) = nodeKey(msg.streamId, msg.toolName, msg.componentId) ?: return emptyList()
    if (node in existingIds) return emptyList()
    val tool = msg.toolName ?: "tool"
    val attrs =
        buildJsonObject {
            put("type", "text")
            put("content", "Streaming $tool…")
        }
    return listOf(CanvasOp("upsert", node, Component("text", node, attrs, emptyList())))
}

fun streamErrorOps(msg: Inbound.StreamErrorMsg): List<CanvasOp> {
    val (node, _) = nodeKey(msg.streamId, msg.toolName) ?: return emptyList()
    val text = msg.error.message ?: msg.error.code ?: "stream error"
    val attrs =
        buildJsonObject {
            put("type", "alert")
            put("variant", "error")
            put("title", "Stream error")
            put("message", text)
        }
    return listOf(CanvasOp("upsert", node, Component("alert", node, attrs, emptyList())))
}
